In [5]:
import pandas as pd

In [6]:
import numpy as np

In [7]:
data=pd.read_csv("C:/Users/rames/Downloads/nba_2013.csv")

In [8]:
data_copy=data.copy()

In [9]:
data.head()

player pos  age bref_team_id   g  gs    mp   fg   fga    fg.  ...  \
0     Quincy Acy  SF   23          TOT  63   0   847   66   141  0.468  ...   
1   Steven Adams   C   20          OKC  81  20  1197   93   185  0.503  ...   
2    Jeff Adrien  PF   27          TOT  53  12   961  143   275  0.520  ...   
3  Arron Afflalo  SG   28          ORL  73  73  2552  464  1011  0.459  ...   
4  Alexis Ajinca   C   25          NOP  56  30   951  136   249  0.546  ...   

   drb  trb  ast  stl  blk  tov   pf   pts     season  season_end  
0  144  216   28   23   26   30  122   171  2013-2014        2013  
1  190  332   43   40   57   71  203   265  2013-2014        2013  
2  204  306   38   24   36   39  108   362  2013-2014        2013  
3  230  262  248   35    3  146  136  1330  2013-2014        2013  
4  183  277   40   23   46   63  187   328  2013-2014        2013  

[5 rows x 31 columns]

In [10]:
data.describe()

age           g          gs           mp          fg  \
count  481.000000  481.000000  481.000000   481.000000  481.000000   
mean    26.509356   53.253638   25.571726  1237.386694  192.881497   
std      4.198265   25.322711   29.658465   897.258840  171.832793   
min     19.000000    1.000000    0.000000     1.000000    0.000000   
25%     23.000000   32.000000    0.000000   388.000000   47.000000   
50%     26.000000   61.000000   10.000000  1141.000000  146.000000   
75%     29.000000   76.000000   54.000000  2016.000000  307.000000   
max     39.000000   83.000000   82.000000  3122.000000  849.000000   

               fga         fg.         x3p        x3pa        x3p.  ...  \
count   481.000000  479.000000  481.000000  481.000000  414.000000  ...   
mean    424.463617    0.436436   39.613306  110.130977    0.285111  ...   
std     368.850833    0.098672   50.855639  132.751732    0.157633  ...   
min       0.000000    0.000000    0.000000    0.000000    0.000000  ...   
25%     110.000000    0.400500    0.000000    3.000000    0.234355  ...   
50%     332.000000    0.438000   16.000000   48.000000    0.330976  ...   
75%     672.000000    0.479500   68.000000  193.000000    0.375000  ...   
max    1688.000000    1.000000  261.000000  615.000000    1.000000  ...   

              orb         drb          trb         ast         stl  \
count  481.000000  481.000000   481.000000  481.000000  481.000000   
mean    55.810811  162.817048   218.627859  112.536383   39.280665   
std     62.101191  145.348116   200.356507  131.019557   34.783590   
min      0.000000    0.000000     0.000000    0.000000    0.000000   
25%     12.000000   43.000000    55.000000   20.000000    9.000000   
50%     35.000000  135.000000   168.000000   65.000000   32.000000   
75%     73.000000  230.000000   310.000000  152.000000   60.000000   
max    440.000000  783.000000  1114.000000  721.000000  191.000000   

              blk         tov          pf          pts  season_end  
count  481.000000  481.000000  481.000000   481.000000       481.0  
mean    24.103950   71.862786  105.869023   516.582121      2013.0  
std     30.875381   62.701690   71.213627   470.422228         0.0  
min      0.000000    0.000000    0.000000     0.000000      2013.0  
25%      4.000000   21.000000   44.000000   115.000000      2013.0  
50%     14.000000   58.000000  104.000000   401.000000      2013.0  
75%     32.000000  108.000000  158.000000   821.000000      2013.0  
max    219.000000  295.000000  273.000000  2593.000000      2013.0  

[8 rows x 27 columns]

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 481 entries, 0 to 480
Data columns (total 31 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   player        481 non-null    object 
 1   pos           481 non-null    object 
 2   age           481 non-null    int64  
 3   bref_team_id  481 non-null    object 
 4   g             481 non-null    int64  
 5   gs            481 non-null    int64  
 6   mp            481 non-null    int64  
 7   fg            481 non-null    int64  
 8   fga           481 non-null    int64  
 9   fg.           479 non-null    float64
 10  x3p           481 non-null    int64  
 11  x3pa          481 non-null    int64  
 12  x3p.          414 non-null    float64
 13  x2p           481 non-null    int64  
 14  x2pa          481 non-null    int64  
 15  x2p.          478 non-null    float64
 16  efg.          479 non-null    float64
 17  ft            481 non-null    int64  
 18  fta           481 non-null    

In [12]:
data.shape

(481, 31)

In [13]:
data.isna().sum()

player           0
pos              0
age              0
bref_team_id     0
g                0
gs               0
mp               0
fg               0
fga              0
fg.              2
x3p              0
x3pa             0
x3p.            67
x2p              0
x2pa             0
x2p.             3
efg.             2
ft               0
fta              0
ft.             20
orb              0
drb              0
trb              0
ast              0
stl              0
blk              0
tov              0
pf               0
pts              0
season           0
season_end       0
dtype: int64

In [14]:
data=data.drop(columns=['x3p.'])

In [15]:
data=data.dropna(axis=0)

In [16]:
data.shape

(460, 30)

In [17]:
data['pos'].unique()

array(['SF', 'C', 'PF', 'SG', 'PG', 'G'], dtype=object)

In [18]:
from sklearn.preprocessing import LabelEncoder

In [19]:
labelencoder = LabelEncoder()

In [20]:
data['pos_num'] = labelencoder.fit_transform(data['pos'])

In [21]:
data=data.drop(columns=['pos'])

In [22]:
data.head()

player  age bref_team_id   g  gs    mp   fg   fga    fg.  x3p  ...  \
0     Quincy Acy   23          TOT  63   0   847   66   141  0.468    4  ...   
1   Steven Adams   20          OKC  81  20  1197   93   185  0.503    0  ...   
2    Jeff Adrien   27          TOT  53  12   961  143   275  0.520    0  ...   
3  Arron Afflalo   28          ORL  73  73  2552  464  1011  0.459  128  ...   
4  Alexis Ajinca   25          NOP  56  30   951  136   249  0.546    0  ...   

   trb  ast  stl  blk  tov   pf   pts     season  season_end  pos_num  
0  216   28   23   26   30  122   171  2013-2014        2013        4  
1  332   43   40   57   71  203   265  2013-2014        2013        0  
2  306   38   24   36   39  108   362  2013-2014        2013        2  
3  262  248   35    3  146  136  1330  2013-2014        2013        5  
4  277   40   23   46   63  187   328  2013-2014        2013        0  

[5 rows x 30 columns]

In [23]:
data['bref_team_id'].unique()

array(['TOT', 'OKC', 'ORL', 'NOP', 'NYK', 'POR', 'MIA', 'MEM', 'BRK',
       'PHI', 'MIL', 'ATL', 'WAS', 'GSW', 'DEN', 'HOU', 'SAS', 'PHO',
       'MIN', 'LAC', 'BOS', 'CLE', 'UTA', 'DET', 'CHA', 'DAL', 'CHI',
       'LAL', 'IND', 'TOR', 'SAC'], dtype=object)

In [24]:
data['bref_team_id_num'] = labelencoder.fit_transform(data['bref_team_id'])

In [25]:
data.head()

player  age bref_team_id   g  gs    mp   fg   fga    fg.  x3p  ...  \
0     Quincy Acy   23          TOT  63   0   847   66   141  0.468    4  ...   
1   Steven Adams   20          OKC  81  20  1197   93   185  0.503    0  ...   
2    Jeff Adrien   27          TOT  53  12   961  143   275  0.520    0  ...   
3  Arron Afflalo   28          ORL  73  73  2552  464  1011  0.459  128  ...   
4  Alexis Ajinca   25          NOP  56  30   951  136   249  0.546    0  ...   

   ast  stl  blk  tov   pf   pts     season  season_end  pos_num  \
0   28   23   26   30  122   171  2013-2014        2013        4   
1   43   40   57   71  203   265  2013-2014        2013        0   
2   38   24   36   39  108   362  2013-2014        2013        2   
3  248   35    3  146  136  1330  2013-2014        2013        5   
4   40   23   46   63  187   328  2013-2014        2013        0   

   bref_team_id_num  
0                28  
1                20  
2                28  
3                21  
4                18  

[5 rows x 31 columns]

In [26]:
data['season'].unique()

array(['2013-2014'], dtype=object)

In [27]:
data=data.drop(columns=['season','bref_team_id'])

In [28]:
data.head()

player  age   g  gs    mp   fg   fga    fg.  x3p  x3pa  ...  trb  \
0     Quincy Acy   23  63   0   847   66   141  0.468    4    15  ...  216   
1   Steven Adams   20  81  20  1197   93   185  0.503    0     0  ...  332   
2    Jeff Adrien   27  53  12   961  143   275  0.520    0     0  ...  306   
3  Arron Afflalo   28  73  73  2552  464  1011  0.459  128   300  ...  262   
4  Alexis Ajinca   25  56  30   951  136   249  0.546    0     1  ...  277   

   ast  stl  blk  tov   pf   pts  season_end  pos_num  bref_team_id_num  
0   28   23   26   30  122   171        2013        4                28  
1   43   40   57   71  203   265        2013        0                20  
2   38   24   36   39  108   362        2013        2                28  
3  248   35    3  146  136  1330        2013        5                21  
4   40   23   46   63  187   328        2013        0                18  

[5 rows x 29 columns]

In [29]:
data['season_end'].unique()

array([2013], dtype=int64)

In [30]:
data=data.drop(columns=['season_end'])

In [31]:
from sklearn import preprocessing

In [32]:
li=list(data.columns)

In [33]:
print(li)

['player', 'age', 'g', 'gs', 'mp', 'fg', 'fga', 'fg.', 'x3p', 'x3pa', 'x2p', 'x2pa', 'x2p.', 'efg.', 'ft', 'fta', 'ft.', 'orb', 'drb', 'trb', 'ast', 'stl', 'blk', 'tov', 'pf', 'pts', 'pos_num', 'bref_team_id_num']


In [34]:
for i in li[1:]:
    p=i
    preprocessing.normalize([np.array(data[p])])

In [35]:
from sklearn.model_selection import train_test_split

In [36]:
y=list(data['pts'])
X=data.drop(columns=["player",'pts'])

In [37]:
y=np.array(y)

In [38]:
y

array([ 171,  265,  362, 1330,  328,   92, 1603,  303,  701,  495,  572,
         38,  477,  564,  810,  436,  525, 2112,   28,  352, 1107,   25,
        401,  280,  930,  112,  240,  170,  150,  660,  558,  738,  625,
        166,  911,  298, 1070,  666,  161,  403, 1249,  436,  911,  217,
        573,    3,   72,  222,  497,  378,  821,  763,    5,  492,  197,
        491, 1042, 1281,  891,  419,  998,    7,  645,  159,   64,   63,
         83,  274,  115,  895, 1089,  587,  878,  136,   43,  225,  485,
        347,  925,  470,  101,  810,  967, 1167,  490,  716,  480,  846,
         71,  257,   68,   47,  527,   25,  911,  338, 1256,  153, 1614,
         33, 1282,  890,  356,  510,   26, 1873,  529,   83,  144, 1394,
        357,  641,   66,  177,   97,  339, 1011, 1791,  717,  286,  202,
        339, 1542, 1095,  511, 1119,  929, 2593,  145, 1560,  401,  214,
       1042, 1096,  416,  164,  970,   19,  630,   68,  425,  638, 1068,
         39,  273,  174,  910,  315,  352,  248,  8

In [39]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.33)

In [40]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression

In [41]:
model=KNeighborsClassifier(n_neighbors=7)
model.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=7)

In [42]:
predictknn=model.predict(X_test)

In [55]:
linear=LinearRegression()
linear.fit(X_train,y_train)
predictionlinear=linear.predict(X_test)
predictionlinear=list(predictionlinear)

In [59]:
for i in range(len(predictionlinear)):
    predictionlinear[i]=int(predictionlinear[i])

In [60]:
predictionlinear=np.array(predictionlinear)

In [61]:
predictionlinear

array([ 256,  224,   34,   18,  349,  268,   73,  472,  720,  205,  182,
        666,  625,  490,  425,  838, 1465,   17,  137,  356,  784,   24,
       1372, 1041,  608,  131,  109,  149,  351,   75,  495,  890,  625,
         28,   11,  114,   27,   70,  285,  158,   13,  429,  101,  159,
        298,  216,  930,  172,  762,  721,  525, 1289,   65,  362,  799,
        891,  944, 1068,  754,   67, 1166, 1478,   62,  201,  106,  273,
          5,   47,  182,  716,  404,  824,  603,  470,  301,  537, 1735,
        457,   37,  548,  990,   96,  573, 1281,  831,  298,  715,  258,
       1602,  703,  170,  350,    0,  338, 1071,  282,  630,  529, 1297,
        480,  448,   91,   81,  212,  519, 1106,   36,  487,  490,  530,
       1614,  629,  274, 1240,  301,   38,  432,  618,  260,  780,  636,
        221, 1006, 2009,   46,  350, 1053, 1583,  871,  436,   10,  844,
        339,   72,  429, 2593,  100,  810,  878,  770,    0, 1131,  532,
        461,    4,  417,  661,  511,   71, 1042,  6

In [87]:
y_test

array([ 257,  225,   35,   19,  349,  268,   74,  472,  720,  206,  183,
        666,  625,  490,  425,  838, 1465,   18,  137,  356,  784,   25,
       1372, 1041,  608,  132,  110,  150,  352,   76,  495,  890,  626,
         29,   12,  115,   28,   71,  286,  159,   14,  430,  102,  159,
        298,  217,  930,  173,  763,  721,  525, 1289,   66,  362,  799,
        891,  944, 1068,  754,   68, 1167, 1478,   63,  202,  107,  273,
          6,   48,  183,  716,  404,  824,  603,  470,  302,  538, 1735,
        457,   38,  548,  990,   97,  573, 1281,  831,  298,  715,  258,
       1603,  703,  171,  350,    0,  338, 1071,  283,  630,  529, 1297,
        480,  448,   92,   82,  213,  520, 1106,   37,  487,  490,  530,
       1614,  629,  274, 1241,  301,   39,  432,  618,  261,  781,  636,
        222, 1007, 2010,   47,  350, 1053, 1583,  871,  436,   11,  844,
        339,   73,  429, 2593,  101,  810,  878,  770,    1, 1131,  532,
        461,    5,  417,  661,  511,   72, 1042,  6

In [67]:
from sklearn.metrics import accuracy_score,confusion_matrix
accuracy_score(y_test,predictionlinear,normalize=False)

89

In [70]:
pip install lightgbm

Note: you may need to restart the kernel to use updated packages.


In [73]:
from lightgbm import LGBMRegressor


In [74]:
gbmmodel = LGBMRegressor(max_depth=6, learning_rate=0.1, n_estimators=500,min_child_weight=100, subsample=1.0,colsample_bytree=0.8,random_state=42, n_jobs=-1)
gbmmodel.fit(X_train,y_train)
gbRegressor=gbmmodel.predict(X_test)

In [76]:
for i in range(len(gbRegressor)):
    gbRegressor[i]=int(gbRegressor[i])

In [82]:
from sklearn import svm

In [94]:
clf=svm.SVR()
clf.fit(X_train,y_train)
pred=clf.predict(X_test)

In [95]:
for i in range(len(pred)):
    pred[i]=int(pred[i])

### From the above regression algorithms we can conclude that Linear prediction has more accuracy rate of 89%